### PR TITLE
Remove lodash deps

### LINKED ISSUE #11726: @babel/core - Get rid of (only 4 instances) lodash to fix security issue
## Feature Request

I would like to remove lodash as a dependency in @babel/core. There has been a very long period where lodash had the chance to fix a prototype pollution vulnerability, and even after long period of security companies advicing privately, the security report was published publicly. And still, the maintainer is openly ignoring the security problem. https://github.com/lodash/lodash/issues/4738

This has lead many dependents to get off lodash, rightfully so. Security should not be ignored for this long, especially with this many weekly downloads.

@babel/core is a pretty huge dependent of lodash, but there are actually only 4 instances where lodash is used:
- escapeRegExp --> ./src/config/pattern-to-regex.js
- escapeRegExp --> ./test/config-chain.js
- sortBy --> ./src/transformation/block-hoist-plugin.js
- cloneDeep --> ./src/transformation/normalize-file.js

Safe to say, lodash is just helpers, and you never actually NEED them, you can just use regular JavaScript or create your own helper if you need it. Reducing dependencies is always a smart security/risk move.

If the babel maintainers agree, I would be happy to try and provide a pull request to replace the instances of lodash in @babel/core so we can get rid of it, and subsequently fix the security vulnerability that so many people face who are using @babel/core. 

Let me know!

- [x] I would like to work on this feature!

### PR DESCRIPTION
**Pull request by hzoo**
Inlining the function in most cases

| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Closes #11726, ref https://github.com/babel/babel/pull/13021
| Any Dependency Changes?  | Lodash
| License                  | MIT

### CODE REVIEW

File: packages/babel-helper-transform-fixture-test-runner/src/index.ts [hunk 1]
```diff
@@ -202,19 +199,17 @@ function run(task) {
...
+    const newOpts = {
+      ast: true,
+      cwd: path.dirname(self.loc),
+      filename: self.loc,
+      filenameRelative: self.filename,
+      sourceFileName: self.filename,
+      sourceType: "script",
+      babelrc: false,
+      inputSourceMap: task.inputSourceMap || undefined,
+      ...opts,
```
Comment (hzoo) [inline packages/babel-helper-transform-fixture-test-runner/src/index.ts:211]: It doesn't seem like we'd need to deep merge any of the options above so should be equivalent (or are all primitives)

File: packages/babel-types/src/builders/builder.ts [hunk 2]
```diff
@@ -23,7 +22,9 @@ export default function builder<T extends t.Node>(
 
     let arg;
     if (i < countArgs) arg = args[i];
-    if (arg === undefined) arg = loClone(field.default);
+    if (arg === undefined) {
+      arg = Array.isArray(field.default) ? [] : field.default;
```
Comment (hzoo) [inline packages/babel-types/src/builders/builder.ts:26]: in my search, all the default options are primitives except for a default of empty array

types

Comment (nicolo-ribaudo) [reply packages/babel-types/src/builders/builder.ts:26]: We should throw if we see a non-empty array or an object, so that we don't accidentally break this assumption in the future.

File: packages/babel-types/src/converters/valueToNode.ts [hunk 3]
```diff
@@ -33,6 +32,10 @@ export default valueToNode as {
   (value: unknown): t.Expression;
 };
 
+function isRegExp(value): boolean {
```
Comment (nicolo-ribaudo) [inline packages/babel-types/src/converters/valueToNode.ts:35]:
```suggestion
function isRegExp(value): value is RegExp {
```

File: Gulpfile.mjs [hunk 4]
```diff
@@ -162,7 +162,9 @@ function generateStandalone() {
         let allList = "";
 
         for (const plugin of pluginConfig) {
-          const camelPlugin = camelCase(plugin);
+          const camelPlugin = plugin.replace(/-([a-z])/g, c =>
```
Comment (nicolo-ribaudo) [inline Gulpfile.mjs:165]:
```suggestion
          const camelPlugin = plugin.replace(/-[a-z]/g, c =>
```

I don't think we need the capturing group here

File: packages/babel-types/src/builders/builder.ts (hunk 2 above)
Comment (hzoo) [reply packages/babel-types/src/builders/builder.ts:26]: ah right, https://lodash.com/docs/#clone seems complicated, maybe we need to check what we do allow instead then given it could be anything?

so string/boolean/null/[]

File: packages/babel-helper-transform-fixture-test-runner/package.json [hunk 5]
```diff
@@ -20,7 +20,6 @@
     "@babel/helper-fixtures": "workspace:^7.13.10",
     "babel-check-duplicated-nodes": "^1.0.0",
     "escape-string-regexp": "condition:BABEL_8_BREAKING ? ^4.0.0 : ",
```
Comment (merceyz) [inline packages/babel-helper-transform-fixture-test-runner/package.json:22]:
```suggestion
```
Can remove this as well now since you removed the file that uses it

File: packages/babel-register/src/escape-regexp.cjs [hunk 6]
```diff
@@ -1,5 +0,0 @@
-"use strict";
-
-module.exports = process.env.BABEL_8_BREAKING
-  ? require("escape-string-regexp")
```
Comment (merceyz) [inline packages/babel-register/src/escape-regexp.cjs:4]: Same here

File: Gulpfile.mjs [hunk 7]
```diff
@@ -8,13 +8,12 @@ import through from "through2";
...
 import newer from "gulp-newer";
 import babel from "gulp-babel";
-import camelCase from "lodash/camelCase.js";
 import fancyLog from "fancy-log";
 import filter from "gulp-filter";
 import revertPath from "gulp-revert-path";
 import gulp from "gulp";
 import { rollup } from "rollup";
-import { babel as rollupBabel } from "@rollup/plugin-babel";
+import _rollupBabel from "@rollup/plugin-babel";
```
Comment (nicolo-ribaudo) [inline Gulpfile.mjs:16]: What is this change needed for? :thinking:

Comment (hzoo) [reply Gulpfile.mjs:16]: Someone weird I had to do for windows? 

It's similar to what we did for "rollup-plugin-dts"?

File: packages/babel-types/src/builders/builder.ts (hunk 2 above)
Comment (nicolo-ribaudo) [reply packages/babel-types/src/builders/builder.ts:26]: Yeah, we can just throw if `Array.isArray(def) ? def.length > 0 : def && typeof def === "object"`

File: Gulpfile.mjs (hunk 7 above)
Comment (nicolo-ribaudo) [reply Gulpfile.mjs:16]: For `rollup-plugin-dts` it's because Node.js cannot recognize its exports, but for `@rollup/plugin-babel` it should work since it does `exports.babel = ...` :thinking: 

Maybe you have a very old Node.js version, from before that Node.js used `cjs-module-lexer`? Otherwise it seems like a Node.js bug.

Comment (hzoo) [reply Gulpfile.mjs:16]: I'll paste the error if that helps? I'm on later node (`$ node -v v14.9.0`)

```
file:///C:/Users/Hen/win-dev/babel/gulpfile.mjs:17
import { babel as rollupBabel } from "@rollup/plugin-babel";
         ^^^^^
SyntaxError: The requested module '@rollup/plugin-babel' is expected to 
be of type CommonJS, which does not support named exports. CommonJS modules can be imported by importing the default export.
For example:
import pkg from '@rollup/plugin-babel';
const { babel: rollupBabel } = pkg;
```

File: packages/babel-helper-transform-fixture-test-runner/package.json (hunk 5 above)
Comment (hzoo) [reply packages/babel-helper-transform-fixture-test-runner/package.json:22]: good catch thanks!!

File: packages/babel-types/src/builders/builder.ts (hunk 2 above)
Comment (hzoo) [reply packages/babel-types/src/builders/builder.ts:26]: Added check in the other file! https://github.com/babel/babel/pull/13057/files#diff-36b14d40dba9fa52476c9907672a7a9e06a1da66dcc135100a918e956aadd04fR283-R289

File: packages/babel-cli/test/index.js [hunk 8]
```diff
@@ -244,7 +246,7 @@ fs.readdirSync(fixtureLoc).forEach(function (binName) {
 
           delete taskOpts.os;
         }
-        merge(opts, taskOpts);
+        opts = { args: [], ...taskOpts };
```
Comment (hzoo) [inline packages/babel-cli/test/index.js:249]: Changed this merge too (the only thing replaced is args, which is always set in our tests given it's babel-cli)

**Review by nicolo-ribaudo — APPROVED** [review]
Thanks!

I love how "inline some deps" results in "more deleted lines than added lines" :joy:

**Review by existentialism — APPROVED** [review]

**Comment by julienw** [comment]
hey @hzoo @nicolo-ribaudo, is it expected that last version of `@babel/cli` still depends on lodash ? See https://github.com/babel/babel/blob/main/packages/babel-cli/package.json#L31 :-)
The CHANGELOG file mentions that it's removed from `@babel/cli` too, hence my question.

Thanks!
